c_plus_plus/src/model/media_player/subtitles/subtitles.cpp

```diff
@@ -0,0 +1,142 @@
+﻿#include "subtitles.h"
+
+#include "srt_file.h"
+
+#include <cmath>
+
+namespace Model
+{
+Subtitles::Subtitles(QObject *parent)
+    : QObject(parent)
+    , m_subtitles({})
+    , m_lastSubtitleIndex(0)
+    , m_currentTime(0)
+    , m_lastSubtitleSignaled(QString())
+    , m_timer(parent)
+{
+    m_timer.setTimerType(Qt::PreciseTimer);
+    m_timer.setInterval(m_timerInterval);
+
+    connect(&m_timer, &QTimer::timeout, [=] { onTimerTimeout(); });
+}
+
+void Subtitles::open(const QString &srtFilePath)
+{
+    m_subtitles = SrtFile::parse(srtFilePath);
+}
+
+void Subtitles::play()
+{
+    if (!m_subtitles.empty())
+    {
+        m_timer.start();
+        subtitleChanged(currentSubtitle(m_currentTime, true));
+    }
+}
+
+void Subtitles::pause()
+{
+    if (!m_subtitles.empty())
+    {
+        m_timer.stop();
+    }
+}
+
+void Subtitles::stop()
+{
+    if (!m_subtitles.empty())
+    {
+        m_timer.stop();
+        reset();
+    }
+}
+
+void Subtitles::setCurrentTime(qint64 time)
+{
+    if (!m_subtitles.empty())
+    {
+        if (m_timer.isActive())
+        {
+            m_timer.start();
+        }
+
+        m_currentTime = time;
+        subtitleChanged(currentSubtitle(m_currentTime, true));
+    }
+}
+
+void Subtitles::onTimerTimeout()
+{
+    m_currentTime += m_timerInterval;
+    QString newSubtitle = currentSubtitle(m_currentTime, false);
+    if (newSubtitle != m_lastSubtitleSignaled)
+    {
+        subtitleChanged(newSubtitle);
+        m_lastSubtitleSignaled = newSubtitle;
+    }
+}
+
+void Subtitles::reset()
+{
+    m_subtitles.clear();
+    m_currentTime = 0;
+    m_lastSubtitleIndex = 0;
+    m_lastSubtitleSignaled = "";
+    subtitleChanged(QString());
+}
+
+QString Subtitles::currentSubtitle(qint64 time, bool manual)
+{
+    QString subtitle = "";
+
+    if (m_subtitles.size() != 0 && time < m_subtitles.back().end)
```

Comment:
I think this function is a bit too long, maybe split it in smaller functions. If you think it's a good idea, maybe we could make a strategy pattern!

Reply:
I slit the linear search and the binary search in 2 functions